File: simulators/de.unistuttgart.iste.sqa.mpw.karasimulator.java/core/src/main/java/de/unistuttgart/kara/viewmodel/impl/KaraGameViewPresenter.java
package de.unistuttgart.kara.viewmodel.impl;

import de.unistuttgart.iste.sqa.mpw.framework.datatypes.Direction;
import de.unistuttgart.iste.sqa.mpw.framework.datatypes.Size;
import de.unistuttgart.iste.sqa.mpw.framework.mpw.Tile;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelCell;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelCellLayer;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.impl.GameViewPresenterBase;
import de.unistuttgart.kara.facade.KaraGame;
import de.unistuttgart.kara.facade.World;
import de.unistuttgart.kara.kara.Leaf;
import de.unistuttgart.kara.kara.Mushroom;
import de.unistuttgart.kara.kara.ReadOnlyKara;
import de.unistuttgart.kara.kara.Tree;
import javafx.beans.property.ReadOnlyListProperty;
import javafx.beans.property.ReadOnlyObjectProperty;
import javafx.beans.value.ChangeListener;

import java.util.Optional;

public class KaraGameViewPresenter extends GameViewPresenterBase {
	private final World world;
	private ChangeListener<Direction> karaDirectionChangeListener;

	public KaraGameViewPresenter(KaraGame game) {
		super(game);
		this.world = game.getWorld();
	}

	@Override
	protected ReadOnlyObjectProperty<Size> getStageSizeFromConcreteStage() {
		return world.getInternalWorld().stageSizeProperty();
	}

	@Override
	protected ReadOnlyListProperty<Tile> getTilesPropertyFromConcreteStage() {
		return world.getInternalWorld().tilesProperty();
	}

	@Override
	protected void onSetTileNodeAtForCell(ViewModelCell cell, Tile tile) {
		configureTreeImageView(cell, tile);
		configureLeafImageView(cell, tile);
		configureMushroomImageView(cell, tile);

		Optional<ReadOnlyKara> karaOptional = findKaraOnTile(tile);
		karaOptional.ifPresent(readOnlyKara -> configureKaraImageView(cell, readOnlyKara));
	}

	private Optional<ReadOnlyKara> findKaraOnTile(Tile tile) {
		return tile.getContents().stream()
				.filter(ReadOnlyKara.class::isInstance)
				.map(ReadOnlyKara.class::cast).findFirst();
	}

	private void configureTreeImageView(ViewModelCell cell, Tile tile) {
		var layer = new ViewModelCellLayer();
		layer.setImageName("Tree");
		refreshTreeLayer(layer, tile);
		cell.getLayers().add(layer);
	}

	private void refreshTreeLayer(ViewModelCellLayer layer, Tile tile) {
		layer.setVisible(tile.getContents().stream().anyMatch(Tree.class::isInstance));
	}

	private void configureLeafImageView(ViewModelCell cell, Tile tile) {
		var layer = new ViewModelCellLayer();
		layer.setImageName("Leaf");
		refreshLeafLayer(layer, tile);
		cell.getLayers().add(layer);
	}

	private void refreshLeafLayer(ViewModelCellLayer layer, Tile tile) {
		layer.setVisible(tile.getContents().stream().anyMatch(Leaf.class::isInstance));
	}

	private void configureMushroomImageView(ViewModelCell cell, Tile tile) {
		var layer = new ViewModelCellLayer();
		layer.setImageName("Mushroom");
		refreshMushroomLayer(layer, tile);
		cell.getLayers().add(layer);
	}

	private void refreshMushroomLayer(ViewModelCellLayer layer, Tile tile) {
		layer.setVisible(tile.getContents().stream().anyMatch(Mushroom.class::isInstance));
	}

	private void configureKaraImageView(ViewModelCell cell, ReadOnlyKara kara) {
		var layer = new ViewModelCellLayer();
		layer.setImageName("Kara");

		addKaraDirectionListener(layer, kara);

		refreshKaraLayer(layer, kara);
		cell.getLayers().add(layer);
	}

	/*
	 * Adds a listener for the change of the direction, to also update the layers if Kara turns left.
	 * Note: Since onSetTileNodeAtForCell() is called every time the contents of a tile changes, Kara might
	 * be configured multiple times. Avoid, that multiple direction listeners are attached.
	 */
	private void addKaraDirectionListener(final ViewModelCellLayer layer, final ReadOnlyKara kara) {
		if (karaDirectionChangeListener != null) {
			kara.directionProperty().removeListener(karaDirectionChangeListener);
		}
		karaDirectionChangeListener = (property, oldValue, newValue) -> {
			runLocked(() -> {
				refreshKaraLayer(layer, kara);
			});
		};
		kara.directionProperty().addListener(karaDirectionChangeListener);
	}

	private void refreshKaraLayer(ViewModelCellLayer layer, ReadOnlyKara kara) {
		layer.setVisible(kara.getCurrentTile() != null);
		if (kara.getDirection() != null) {
			layer.setRotation(getRotationForDirection(kara.getDirection()));
		}
	}

}

File: core/src/impl/KaraGameViewPresenter.cpp
#include "KaraGameViewPresenter.h"

#include "World.h"
#include "GameWorld.h"
#include "GameLog.h"
#include "CommandStack.h"
#include "GamePerformance.h"
#include "ReadOnlyKara.h"
#include "Actor.h"

#include "ViewModelCellLayer.h"

#include "CollectionHelpers.hpp"
#include <utility>

using namespace mpw;
using namespace kara;
using namespace collectionhelpers;
using namespace framework;

namespace viewmodel {

KaraGameViewPresenter::KaraGameViewPresenter(std::shared_ptr<kara::KaraGame> game)
        : inherited(game), game(std::move(game)) {
}

const ObservablePrimitiveProperty<mpw::Size>& KaraGameViewPresenter::getStageSizeFromConcreteStage() {
    return this->game->getWorld()->getInternalWorld()->stageSizeProperty();
}

const framework::ObservableListProperty<mpw::Tile>& KaraGameViewPresenter::getTilesPropertyFromConcreteStage() {
    return game->getWorld()->getInternalWorld()->tilesProperty();
}

void KaraGameViewPresenter::onSetTileNodeAtForCell(ViewModelCell& cell, const mpw::Tile& tile) {
    configureTreeImageView(cell, tile);
    configureMushroomImageView(cell, tile);
    configureLeafImageView(cell, tile);
    auto karaOptional = findKaraOnTile(tile);
    if (karaOptional.has_value()) {
        configureKaraImageView(cell, *karaOptional.value());
    }
}

std::optional<std::shared_ptr<const kara::ReadOnlyKara>> KaraGameViewPresenter::findKaraOnTile(const Tile& tile) {
    const auto listOfKaras = type_select<kara::ReadOnlyKara>(tile.getContents());
    if (listOfKaras.empty()) {
        return std::nullopt;
    }
    return listOfKaras.front();
}

void KaraGameViewPresenter::configureTreeImageView(ViewModelCell& cell, const Tile& tile) {
    auto layer = std::make_shared<ViewModelCellLayer>();
    layer->setImageName("Tree");
    refreshTreeLayer(*layer, tile);

    cell.addToLayers(layer);
}

void KaraGameViewPresenter::refreshTreeLayer(ViewModelCellLayer& layer, const Tile& tile) {
    layer.setVisible(!getTreesOfTile(tile).empty());
}

void KaraGameViewPresenter::configureLeafImageView(ViewModelCell& cell, const Tile& tile) {
    auto layer = std::make_shared<ViewModelCellLayer>();
    layer->setImageName("Leaf");
    refreshLeafLayer(*layer, tile);

    cell.addToLayers(layer);
}

void KaraGameViewPresenter::refreshLeafLayer(ViewModelCellLayer& layer, const Tile& tile) {
    layer.setVisible(!getLeafsOfTile(tile).empty());
}

void KaraGameViewPresenter::configureMushroomImageView(ViewModelCell& cell, const Tile& tile) {
    auto layer = std::make_shared<ViewModelCellLayer>();
    layer->setImageName("Mushroom");
    refreshMushroomLayer(*layer, tile);

    cell.addToLayers(layer);
}

void KaraGameViewPresenter::refreshMushroomLayer(ViewModelCellLayer& layer, const Tile& tile) {
    layer.setVisible(!getMushroomsOfTile(tile).empty());
}

void KaraGameViewPresenter::configureKaraImageView(ViewModelCell& cell, const ReadOnlyKara& kara) {
    auto layer = std::make_shared<ViewModelCellLayer>();
    layer->setImageName("Kara");

    addKaraDirectionListener(layer, kara);
    refreshKaraLayer(*layer, kara);

    cell.addToLayers(layer);
}

/*
 * Adds a listener for the change of the direction, to also update the layers if Kara turns left.
 * Note: Since onSetTileNodeAtForCell() is called every time the contents of a tile changes, Kara might
 * be configured multiple times. Avoid, that multiple direction listeners are attached.
 */
void KaraGameViewPresenter::addKaraDirectionListener(std::shared_ptr<ViewModelCellLayer> karaLayer, const ReadOnlyKara& kara) {
    if (karaDirectionChangeListenerId.has_value()) {
        kara.directionProperty().removeListener(karaDirectionChangeListenerId.value());
    }
    karaDirectionChangeListenerId = kara.directionProperty().addListener(
            [this, &kara, karaLayer](Direction oldValue, Direction newValue) {
                auto lock = getSemaphore().lock();
                refreshKaraLayer(*karaLayer, kara);
            });
}

void KaraGameViewPresenter::refreshKaraLayer(ViewModelCellLayer& layer, const kara::ReadOnlyKara& kara) {
    layer.setVisible(kara.getCurrentTile() != nullptr);
    if (kara.getCurrentTile() == nullptr) {
        return;
    }
    layer.setRotation(getRotationForDirection(kara.getDirection()));
}

std::list<std::shared_ptr<const kara::Tree>> KaraGameViewPresenter::getTreesOfTile(const Tile& tile) {
    return type_select<kara::Tree>(tile.getContents());
}

std::list<std::shared_ptr<const kara::Leaf>> KaraGameViewPresenter::getLeafsOfTile(const Tile& tile) {
    return type_select<kara::Leaf>(tile.getContents());
}

std::list<std::shared_ptr<const kara::Mushroom>> KaraGameViewPresenter::getMushroomsOfTile(const Tile& tile) {
    return type_select<kara::Mushroom>(tile.getContents());
}

}
